server/player/player.go

```diff
@@ -1160,6 +1197,75 @@ func (p *Player) Jump() {
 	}
 }
 
+// Sleep makes the player sleep at the given position. If the position does not map to a bed (specifically the head side),
+// the player will not sleep.
+func (p *Player) Sleep(pos cube.Pos) {
+	ctx, sendReminder := event.C(), true
+	if p.Handler().HandleSleep(ctx, &sendReminder); ctx.Cancelled() {
+		return
+	}
+
+	w := p.World()
+	if b, ok := w.Block(pos).(block.Bed); ok {
+		if b.User != nil {
+			// The player cannot sleep here.
+			return
+		}
+		b.User = p
+		w.SetBlock(pos, b, nil)
+	}
+
+	w.SetRequiredSleepDuration(time.Second * 5)
+	if sendReminder {
+		w.BroadcastSleepingReminder(p)
+	}
+
+	vec := pos.Vec3Middle().Add(mgl64.Vec3{0, 0.5625})
+
+	p.pos.Store(&vec)
+	p.sleeping.Store(true)
+	p.sleepPos.Store(&pos)
+
+	w.BroadcastSleepingIndicator()
+	p.updateState()
+}
+
+// Wake forces the player out of bed if they are sleeping.
+func (p *Player) Wake() {
+	if !p.sleeping.CompareAndSwap(true, false) {
+		return
+	}
+
+	w := p.World()
+	w.SetRequiredSleepDuration(0)
```

Comment:
Is this correct? Wake() can be called from many places such as a player leaving, a bed being destroyed etc.